### ISSUE TITLE
ipv4 works with metallb layer 2, but ipv6 doesn't

### ISSUE THREAD

**Issue by cortopy** [cilium/cilium#10623]
## Bug report

**General Information**

I'm starting with Cilium and after testing many configurations I'm able to have LoadBalancer services working for ipv4. However, in a dual stack setup, even though a service using a ipv4 address works, a service with a ipv6 pointing to the same backend doesn't.

At this stage I don't know if this is a bug or a failed config on my part. But this is the most important things of my debugging.

I have a hello-world app with the same service deployed three times to make sure metallb is assigning the IPs from the pool (see details below)

```
root@master0:~# kubectl get services
NAME           TYPE           CLUSTER-IP       EXTERNAL-IP           PORT(S)          AGE
hello-node     LoadBalancer   10.108.78.191    88.198.105.214        8080:31893/TCP   25s
hello-node-2   LoadBalancer   10.105.249.222   2a01:4f8:c01f:84::    8080:30432/TCP   14s
hello-node-3   LoadBalancer   10.108.91.108    2a01:4f8:c01f:84::1   8080:32067/TCP   5s
kubernetes     ClusterIP      10.96.0.1        <none>                443/TCP          10m
``` 

I've also checked that cilium has created the routes accordingly

```
root@master0:~# kubectl -n kube-system exec -it cilium-7kxjh -- cilium service list
ID   Frontend                     Service Type   Backend                   
1    10.96.0.10:53                ClusterIP      1 => 192.168.0.18:53      
                                                 2 => 192.168.0.11:53      
2    10.96.0.10:9153              ClusterIP      1 => 192.168.0.18:9153    
                                                 2 => 192.168.0.11:9153    
3    10.102.99.237:9189           ClusterIP      1 => 192.168.1.191:9189   
4    10.107.92.236:9189           ClusterIP      1 => 192.168.1.174:9189   
5    10.96.0.1:443                ClusterIP      1 => 10.0.2.1:6443        
6    10.108.78.191:8080           ClusterIP      1 => 192.168.1.131:8080   
7    0.0.0.0:31893                NodePort       1 => 192.168.1.131:8080   
8    49.12.32.101:31893           NodePort       1 => 192.168.1.131:8080   
9    192.168.1.248:31893          NodePort       1 => 192.168.1.131:8080   
10   88.198.105.214:8080          LoadBalancer   1 => 192.168.1.131:8080   
11   10.105.249.222:8080          ClusterIP      1 => 192.168.1.131:8080   
12   0.0.0.0:30432                NodePort       1 => 192.168.1.131:8080   
13   49.12.32.101:30432           NodePort       1 => 192.168.1.131:8080   
14   192.168.1.248:30432          NodePort       1 => 192.168.1.131:8080   
15   [2a01:4f8:c01f:84::]:8080    LoadBalancer   1 => 192.168.1.131:8080   
16   10.108.91.108:8080           ClusterIP      1 => 192.168.1.131:8080   
17   0.0.0.0:32067                NodePort       1 => 192.168.1.131:8080   
18   49.12.32.101:32067           NodePort       1 => 192.168.1.131:8080   
19   192.168.1.248:32067          NodePort       1 => 192.168.1.131:8080   
20   [2a01:4f8:c01f:84::1]:8080   LoadBalancer   1 => 192.168.1.131:8080   
```

When I make a request to any of the NodePort(s) or the ipv4 LoadBalancer `88.198.105.214:8080`, I get the "hello-world" response as expected.

However when doing `curl [2a01:4f8:c01f:84::1]:8080` the connection just hangs.

The tcdump of the two connections shows that while the ipv4 endpoint directs the packets ok, in the case of ipv6 there are no response packets

This is the tcpdump for ipv4:
```
root@worker0:~# tcpdump -i eth0 port 8080
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
19:26:14.795515 IP static.45.157.217.95.clients.your-server.de.54140 > static.88-198-105-214.clients.your-server.de.http-alt: Flags [S], seq 3163642722, win 64240, options [mss 1460,sackOK,TS val 2200144489 ecr 0,nop,wscale 7], length 0
19:26:14.797909 IP static.88-198-105-214.clients.your-server.de.http-alt > static.45.157.217.95.clients.your-server.de.54140: Flags [S.], seq 963664530, ack 3163642723, win 65160, options [mss 1460,sackOK,TS val 80891883 ecr 2200144489,nop,wscale 7], length 0
19:26:14.823167 IP static.45.157.217.95.clients.your-server.de.54140 > static.88-198-105-214.clients.your-server.de.http-alt: Flags [.], ack 1, win 502, options [nop,nop,TS val 2200144516 ecr 80891883], length 0
19:26:14.823293 IP static.45.157.217.95.clients.your-server.de.54140 > static.88-198-105-214.clients.your-server.de.http-alt: Flags [P.], seq 1:84, ack 1, win 502, options [nop,nop,TS val 2200144517 ecr 80891883], length 83: HTTP: GET / HTTP/1.1
19:26:14.823467 IP static.88-198-105-214.clients.your-server.de.http-alt > static.45.157.217.95.clients.your-server.de.54140: Flags [.], ack 84, win 509, options [nop,nop,TS val 80891909 ecr 2200144517], length 0
19:26:14.827802 IP static.88-198-105-214.clients.your-server.de.http-alt > static.45.157.217.95.clients.your-server.de.54140: Flags [P.], seq 1:131, ack 84, win 509, options [nop,nop,TS val 80891913 ecr 2200144517], length 130: HTTP: HTTP/1.1 200 OK
19:26:14.852886 IP static.45.157.217.95.clients.your-server.de.54140 > static.88-198-105-214.clients.your-server.de.http-alt: Flags [.], ack 131, win 501, options [nop,nop,TS val 2200144546 ecr 80891913], length 0
19:26:14.853298 IP static.45.157.217.95.clients.your-server.de.54140 > static.88-198-105-214.clients.your-server.de.http-alt: Flags [F.], seq 84, ack 131, win 501, options [nop,nop,TS val 2200144547 ecr 80891913], length 0
19:26:14.854462 IP static.88-198-105-214.clients.your-server.de.http-alt > static.45.157.217.95.clients.your-server.de.54140: Flags [F.], seq 131, ack 85, win 509, options [nop,nop,TS val 80891940 ecr 2200144547], length 0
19:26:14.879622 IP static.45.157.217.95.clients.your-server.de.54140 > static.88-198-105-214.clients.your-server.de.http-alt: Flags [.], ack 132, win 501, options [nop,nop,TS val 2200144573 ecr 80891940], length 0
```
And this is the tcpdump for ipv6
```
root@worker0:~# tcpdump -i eth0 port 8080
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
19:22:12.155736 IP6 2a01:4f9:c010:74f5::1.58440 > 2a01:4f8:c01f:84::1.http-alt: Flags [S], seq 3531473308, win 64800, options [mss 1440,sackOK,TS val 1381038665 ecr 0,nop,wscale 7], length 0
19:22:13.157403 IP6 2a01:4f9:c010:74f5::1.58440 > 2a01:4f8:c01f:84::1.http-alt: Flags [S], seq 3531473308, win 64800, options [mss 1440,sackOK,TS val 1381039667 ecr 0,nop,wscale 7], length 0
19:22:15.173268 IP6 2a01:4f9:c010:74f5::1.58440 > 2a01:4f8:c01f:84::1.http-alt: Flags [S], seq 3531473308, win 64800, options [mss 1440,sackOK,TS val 1381041683 ecr 0,nop,wscale 7], length 0
19:22:19.205179 IP6 2a01:4f9:c010:74f5::1.58440 > 2a01:4f8:c01f:84::1.http-alt: Flags [S], seq 3531473308, win 64800, options [mss 1440,sackOK,TS val 1381045715 ecr 0,nop,wscale 7], length 0
```

- Cilium version 1.7.1
- Kernel version 5.3.0
- Orchestration system version in use: kubectl 1.17.4
- Link to relevant artifacts (policies, deployments scripts, ...)
- Upload a system dump (see attached)

**How to reproduce the issue**

I have a new k8s cluster with cilium + metallb and without kube-proxy

The cluster is init with kubeadm
```
kubeadm init --skip-phases=addon/kube-proxy  --config kubeadm.yml
```

where the config is like this:
```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: InitConfiguration
localAPIEndpoint:
  advertiseAddress: "10.0.2.1"
  bindPort: 6443
---
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
apiServer:
  certSANs:
    - 10.0.0.1
clusterName: kubernetes
controlPlaneEndpoint: control-plane:6443
controllerManager:
  extraArgs:
    allocate-node-cidrs: "true"
kubernetesVersion: v1.17.4
networking:
  podSubnet: 192.168.0.0/16
```
Metallb is installed using helm and this config
```yaml
configInline:
  address-pools:
  - name: default
    protocol: layer2
    addresses:
    - 88.198.105.214/32
    - 2a01:4f8:c01f:84::1/64
```

Cillium is also installed using helm and these values
```yaml
global:
  k8sServiceHost: control-plane
  k8sServicePort: 6443
  tag: v1.7.1
  ipv4:
    enabled: True
  ipv6:
    enabled: True
  bpf:
    waitForMount: true
  kubeProxyReplacement: "strict"
```

And then, I create a simple deployment [hello-world application](https://kubernetes.io/docs/tutorials/hello-minikube/) using:

```
kubectl create deployment hello-node --image=gcr.io/hello-minikube-zero-install/hello-node
```

and expose with the following service, created three times to make sure that metallb is assigning the IPs from the pool correctly

This is the service:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: hello-node
spec:
  type: LoadBalancer
  selector:
    app: hello-node
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 8080
```
in my /etc/sysctl.conf I have
```
net.ipv6.conf.all.forwarding=1
net.ipv6.conf.all.disable_ipv6=0
net.ipv6.conf.default.disable_ipv6=0
```

[cilium-sysdump-20200318-201710.zip](https://github.com/cilium/cilium/files/4350921/cilium-sysdump-20200318-201710.zip)

**Comment by brb** [comment]
I think the problem is that you are accessing a service via IPv6 address while the backend addr is IPv4.

Have you tried to set `ipFamily: IPv6` in the Service object?

**Comment by cortopy** [comment]
Thanks for that one @brb, I hadn't come across `ipFamily` before as this is the first dual stack I'm trying to build. This has pointed me in the right direction. I now suspect this is not an issue with Cilium

Since I first created this issue, I've been learning more about Cilium. Although I'm still getting there, I recently discovered `cilium monitor` to debug this case. Requests to the IPv4 ip reach cilium, whereas those to the IPv6 do not get logged. 

This has led me to believe that the issue lies with the configuration of the cluster itself and not Cilium. Although I'm not entirely sure this is the case, I'm going to close this issue with this comment.

If someone wants to explore this further, this [guide](https://v1-16.docs.kubernetes.io/docs/concepts/services-networking/dual-stack/) has helped, although I couldn't get the cluster up and running with those settings
